### PR TITLE
Make CVaR work with exact expectation values

### PR DESCRIPTION
**Pull request by laurgao**
From Jira: Guoming suggested that CVaR could actually work without the need for sampling, which should be faster.

Here's his explanation:

You first compute the probability p(x) for each n-bitstring x  from the wavefunction, i.e. p(x)=|amplitude of x|^2. Then sort all the n-bitstrings according to their energies, say, x_1,x_2,... such that E(x_1)<= E(x_2) <= …. Then compute the cumulative sum of the probability s_k = p(x_1) + … + p(x_k) until s_k >= alpha. Then compute the conditioned mean value of E(x_j) for 1<=j<=k , where x_j has probability p(x_j) for j<k and alpha-s_{k-1} for j=k . This value is the limit of CVaR for a large number of samples.

### CODE REVIEW

File: tests/zquantum/qaoa/estimators/cvar_estimator_test.py
Comment (mstechly) [inline tests/zquantum/qaoa/estimators/cvar_estimator_test.py:41]: I would just replace this whole fixture by:

```
def backend(self):
    return MockQuantumBackend()
```

I don't think we need any of this custom code anymore.

File: tests/zquantum/qaoa/estimators/cvar_estimator_test.py
Comment (mstechly) [inline tests/zquantum/qaoa/estimators/cvar_estimator_test.py:118]: Actually, I think that for calculating with exact expectation values you don't even need `pytest.approx`.

File: tests/zquantum/qaoa/estimators/cvar_estimator_test.py
Comment (mstechly) [inline tests/zquantum/qaoa/estimators/cvar_estimator_test.py:97]: Instead of having this test, what about change the `estimator` fixture to:

```
@pytest.fixture(
    params=[
        {
            "alpha": 0.2,
            "use_exact_expectation_values": False,
        },
        {
            "alpha": 0.2,
            "use_exact_expectation_values": True,
        },

    ]
)
def estimator(request):
    return CvarEstimator(**request.param)
```
?
It would require some more test cases, but I think we can limit ourselves to 3 values of alpha.

**Review by mstechly — COMMENTED** [review]
Code-wise looks good, left some comments.
Though I'd like @Guoming-Wang to look over the logic :)

File: tests/zquantum/qaoa/estimators/cvar_estimator_test.py
Comment (laurgao) [reply tests/zquantum/qaoa/estimators/cvar_estimator_test.py:41]: I got rid of `custom_run_circuit_and_measure` but I think we still need `custom_get_wavefunction` (it's not implemented in MockQuantumBackend)

Comment (mstechly) [reply tests/zquantum/qaoa/estimators/cvar_estimator_test.py:64]: True, the way it's set up it won't work.
Then perhaps we need to use `SymbolicSimulator` instead of `MockQuantumBackend` 🤷 
It's also in core.

**Review by mstechly — APPROVED** [review]
Looks good!
Please wait for Guoming's approval until merging though :)

File: src/python/zquantum/qaoa/estimators/_cvar_estimator.py
Comment (Guoming-Wang) [inline src/python/zquantum/qaoa/estimators/_cvar_estimator.py:113]: I am not familiar with wavefunction.amplitudes. Please make sure that the -1 is needed.

**Review by Guoming-Wang — APPROVED** [review]
Looks great to me!

**Pull request merged**
